Add Release note for Airflow 2.3.0 (#23260)
mistercrunch/Airflow: RELEASE_NOTES.rst

- Added windows extensions (#16110)
- Showing approximate time until next dag_run in Airflow  (#20273)
- Extend config window on UI (#20052)
- Add show dag dependencies feature to CLI (#19985)
- Add cli command for 'airflow dags reserialize` (#19471)
- Add missing description field to Pool schema(REST API) (#19841)
- Introduce DagRun action to change state to queued. (#19353)
- Add DAG run details page (#19705)
- Add role export/import to cli tools (#18916)
- Adding ``dag_id_pattern`` parameter to the ``/dags`` endpoint (#18924)

Improvements
""""""""""""

- Show schedule_interval/timetable description in UI (#16931)
- Added column duration to DAG runs view (#19482)
- Enable use of custom conn extra fields without prefix (#22607)
- Initialize finished counter at zero (#23080)
- Improve logging of optional provider features messages (#23037)
- Meaningful error mssage in resolve_template_files (#23027)
- Update ImportError items instead of deleting and recreating them (#22928)
- Add option ``--skip-init`` to db reset command (#22989)
- Support importing connections from files with ".yml" extension (#22872)
- Support glob syntax in ``.airflowignore`` files (#21392) (#22051)

> - Hide pagination when data is a single page (#22963)

Author: ```suggestion
```

  IFrame. By default Airflow could not be embedded in an IFrame.

  In Airflow 2.2.5 there was a bug introduced that made it impossible to disable Airflow to
  work in IFrame. No matter what the configuration was set, it was possible to embed Airflow in an IFrame.

  Airflow 2.3.0 restores the original meaning to the parameter. If you set it to "true" (default) Airflow
  can be embedded in an IFrame (no header is added), but when you set it to "false" the header is added
  and Airflow cannot be embedded in an IFrame. (#23222)

Features
^^^^^^^^

- Add dynamic task mapping (`AIP-42 <https://github.com/apache/airflow/pulls?q=is%3Apr+is%3Amerged+label%3AAIP-42>`_)
- New Grid View replaces Tree View (#18675)
- Templated ``requirements.txt`` in Python Operators (#17349)
- Allow reuse of decorated tasks (#22941)
- Move the database configuration to a new section (#22284)
- Add ``SmoothOperator`` (#22813)
- make operator's ``execution_timeout`` configurable (#22389)
- Events Timetable (#22332)
- Support dag serialization with custom ``ti_deps`` rules (#22698)
- Support log download in task log view (#22804)

> - Add summary of runs to dag details (#22766)

Reviewer: This was part of Tree -> Grid, we shouldn't call it out separately

> - Add summary of runs to dag details (#22766)

Author: ```suggestion
```

- Added windows extensions (#16110)
- Showing approximate time until next dag_run in Airflow  (#20273)
- Extend config window on UI (#20052)
- Add show dag dependencies feature to CLI (#19985)
- Add cli command for 'airflow dags reserialize` (#19471)
- Add missing description field to Pool schema(REST API) (#19841)
- Introduce DagRun action to change state to queued. (#19353)
- Add DAG run details page (#19705)
- Add role export/import to cli tools (#18916)
- Adding ``dag_id_pattern`` parameter to the ``/dags`` endpoint (#18924)

Improvements
^^^^^^^^^^^^

- Show schedule_interval/timetable description in UI (#16931)
- Added column duration to DAG runs view (#19482)
- Enable use of custom conn extra fields without prefix (#22607)
- Initialize finished counter at zero (#23080)
- Improve logging of optional provider features messages (#23037)
- Meaningful error mssage in resolve_template_files (#23027)
- Update ImportError items instead of deleting and recreating them (#22928)
- Add option ``--skip-init`` to db reset command (#22989)
- Support importing connections from files with ".yml" extension (#22872)
- Support glob syntax in ``.airflowignore`` files (#21392) (#22051)

> - Hide pagination when data is a single page (#22963)

Reviewer: This was also part of #18675. Let's remove it

.. towncrier release notes start

Airflow 2.3.0 (2022-04-29)
--------------------------

> 

Author: ```suggestion

For production docker image related changes, see `Docker Image Changelog <https://github.com/apache/airflow/blob/main/docs/docker-stack/changelog.rst#airflow-23>`_

```
@potiuk WDYT

Reviewer 2: ```suggestion
For production docker image related changes, see the `Docker Image Changelog <https://airflow.apache.org/docs/docker-stack/changelog.html>`_.
```

Reviewer 3: SGTM :+1:

Reviewer 3: Yeah. The released in docs is better :)

Reviewer 3: BTW. I am working on and update to the image build scripts to make some parameters a bit clearer - for refreshing the image and to re-add building the image for dockerhub with multi-platform image. 

It has no impact on airflow package/code just the scripts to build image and the Dockerfile itself: https://github.com/apache/airflow/pull/23267

I hope we will be able to merge it before cutting the branch :)

Reviewer 3: I think it shoudl build now - pushed the version that has a good chance to work.

> For production docker image related changes, see the `Docker Image Changelog <https://airflow.apache.org/docs/docker-stack/changelog.html>`_.

Reviewer 2: ```suggestion
For production docker image related changes, see the `Docker Image Changelog <https://airflow.apache.org/docs/docker-stack/changelog.html>`_.

```

Significant Changes
^^^^^^^^^^^^^^^^^^^

> Passing ``execution_date`` to ``XCom.set()``\ , ``XCom.clear()``\ , ``XCom.get_one()``\ , and ``XCom.get_many()`` is deprecated (#19825)

Reviewer 4: ```suggestion
Passing ``execution_date`` to ``XCom.set()``, ``XCom.clear()`` , ``XCom.get_one()`` , and ``XCom.get_many()`` is deprecated (#19825)
```

""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
Continuing the effort to bind TaskInstance to a DagRun, XCom entries are now also tied to a DagRun. Use the ``run_id`` argument to specify the DagRun instead.

> Task log templates are now read from the metadatabase instead of ``airflow.cfg`` (#20165)

Author: ```suggestion
Task log templates are now read from the metadata database instead of ``airflow.cfg`` (#20165)
```

"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
  Previously, a task’s log is dynamically rendered from the ``[core] log_filename_template`` and ``[elasticsearch] log_id_template`` config values at runtime. This resulted in unfortunate characteristics, e.g. it is impractical to modify the config value after an Airflow instance is running for a while, since all existing task logs have be saved under the previous format and cannot be found with the new config value.

>   A new ``log_template`` table is introduced to solve this problem. This table is synchronised with the aforementioned config values every time Airflow starts, and a new field ``log_template_id`` is added to every DAG run to point to the format used by tasks (\ ``NULL`` indicates the first ever entry for compatibility).

Reviewer 4: ```suggestion
  A new ``log_template`` table is introduced to solve this problem. This table is synchronised with the aforementioned config values every time Airflow starts, and a new field ``log_template_id`` is added to every DAG run to point to the format used by tasks (``NULL`` indicates the first ever entry for compatibility).
```

Minimum kubernetes version bumped from ``3.0.0`` to ``21.7.0`` (#20759)
"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""

  No change in behavior is expected.  This was necessary in order to take advantage of a `bugfix <https://github.com/kubernetes-client/python-base/commit/70b78cd8488068c014b6d762a0c8d358273865b4>`_ concerning refreshing of Kubernetes API tokens with EKS, which enabled the removal of some `workaround code <https://github.com/apache/airflow/pull/20759>`_.

> XCom now define ``run_id`` instead of ``execution_date`` (#20975)

Author: ```suggestion
XCom now defined by ``run_id`` instead of ``execution_date`` (#20975)
```

  Note that Airflow’s metadatabase definition on both the database and ORM levels are considered implementation detail without strict backward compatibility guarantees.

Non-JSON-serializable params deprecated (#21135).
"""""""""""""""""""""""""""""""""""""""""""""""""

  It was previously possible to use dag or task param defaults that were not JSON-serializable.

  For example this worked previously:

  .. code-block:: python

     @dag.task(params={"a": {1, 2, 3}, "b": pendulum.now()})
     def datetime_param(value):
         print(value)

     datetime_param("{{ params.a }} | {{ params.b }}")

  Note the use of ``set`` and ``datetime`` types, which are not JSON-serializable.  This behavior is problematic because to override these values in a dag run conf, you must use JSON, which could make these params non-overridable.  Another problem is that the support for param validation assumes JSON.  Use of non-JSON-serializable params will be removed in Airflow 3.0 and until then, use of them will produce a warning at parse time.

> You have to use ``postgresql://`` instead of ``postgres://`` in ``sql_alchemy_conn`` for SQLAlchemy 1.4.0+ (#21205)

Author: ```suggestion
You must use ``postgresql://`` instead of ``postgres://`` in ``sql_alchemy_conn`` for SQLAlchemy 1.4.0+ (#21205)
```

  .. code-block::

     >       raise exc.NoSuchModuleError(
                 "Can't load plugin: %s:%s" % (self.group, name)
             )
     E       sqlalchemy.exc.NoSuchModuleError: Can't load plugin: sqlalchemy.dialects:postgres

  If you cannot change the scheme of your URL immediately, Airflow continues to work with SQLAlchemy
  1.3 and you can downgrade SQLAlchemy, but we recommend updating the scheme.
  Details in the `SQLAlchemy Changelog <https://docs.sqlalchemy.org/en/14/changelog/changelog_14.html#change-3687655465c25a39b968b4f5f6e9170b>`_.

``auth_backends`` replaces ``auth_backend`` configuration setting (#21472)
""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""

  Previously, only one backend was used to authorize use of the REST API. In 2.3 this was changed to support multiple backends, separated by whitespace. Each will be tried in turn until a successful response is returned.

  This setting is also used for the deprecated experimental API, which only uses the first option even if multiple are given.

``airflow.models.base.Operator`` is removed (#21505)
""""""""""""""""""""""""""""""""""""""""""""""""""""

>   Previously, there was an empty class ``airflow.models.base.Operator`` for “type hinting”. This class was never really useful for anything (everything it did could be done better with ``airflow.models.baseoperator.BaseOperator``\ ), and has been removed. If you are relying on the class’s existence, use ``BaseOperator`` (for concrete operators), ``airflow.models.abstractoperator.AbstractOperator`` (the base class of both ``BaseOperator`` and the AIP-42 ``MappedOperator``\ ), or ``airflow.models.operator.Operator`` (a union type ``BaseOperator | MappedOperator`` for type annotation).

Reviewer 4: ```suggestion
  Previously, there was an empty class ``airflow.models.base.Operator`` for “type hinting”. This class was never really useful for anything (everything it did could be done better with ``airflow.models.baseoperator.BaseOperator``), and has been removed. If you are relying on the class’s existence, use ``BaseOperator`` (for concrete operators), ``airflow.models.abstractoperator.AbstractOperator`` (the base class of both ``BaseOperator`` and the AIP-42 ``MappedOperator``), or ``airflow.models.operator.Operator`` (a union type ``BaseOperator | MappedOperator`` for type annotation).
```

Zip files in the DAGs folder can no longer have a ``.py`` extension (#21538)
""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""

  It was previously possible to have any extension for zip files in the DAGs folder. Now ``.py`` files are going to be loaded as modules without checking whether it is a zip file, as it leads to less IO. If a ``.py`` file in the DAGs folder is a zip compressed file, parsing it will fail with an exception.

``auth_backends`` includes session (#21640)
"""""""""""""""""""""""""""""""""""""""""""

>   To allow the Airflow UI to use the API, the previous default authorization backend ``airflow.api.auth.backend.deny_all`` is changed to ``airflow.api.auth.backend.session``\ , and this is automatically added to the list of API authorization backends if a non-default value is set.

Reviewer 4: ```suggestion
  To allow the Airflow UI to use the API, the previous default authorization backend ``airflow.api.auth.backend.deny_all`` is changed to ``airflow.api.auth.backend.session``, and this is automatically added to the list of API authorization backends if a non-default value is set.
```

Default templates for log filenames and elasticsearch log_id changed (#21734)
"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""

  In order to support Dynamic Task Mapping the default templates for per-task instance logging has changed. If your config contains the old default values they will be upgraded-in-place.

  If you are happy with the new config values you should *remove* the setting in ``airflow.cfg`` and let the default value be used. Old default values were:

>   * ``[core] log_filename_template``\ : ``{{ ti.dag_id }}/{{ ti.task_id }}/{{ ts }}/{{ try_number }}.log``
>   * ``[elasticsearch] log_id_template``\ : ``{dag_id}-{task_id}-{execution_date}-{try_number}``

Reviewer 4: ```suggestion
  * ``[core] log_filename_template``: ``{{ ti.dag_id }}/{{ ti.task_id }}/{{ ts }}/{{ try_number }}.log``
  * ``[elasticsearch] log_id_template``: ``{dag_id}-{task_id}-{execution_date}-{try_number}``
```

The webserver.X_FRAME_ENABLED configuration works according to description now (#23222).
""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""

  In Airflow 2.0.0 - 2.2.4 the webserver.X_FRAME_ENABLED parameter worked the opposite of its description,
  setting the value to "true" caused "X-Frame-Options" header to "DENY" (not allowing Airflow to be used
  in an IFrame). When you set it to "false", the header was not added, so Airflow could be embedded in an
  IFrame. By default Airflow could not be embedded in an IFrame.

  In Airflow 2.2.5 there was a bug introduced that made it impossible to disable Airflow to
  work in IFrame. No matter what the configuration was set, it was possible to embed Airflow in an IFrame.

  Airflow 2.3.0 restores the original meaning to the parameter. If you set it to "true" (default) Airflow
  can be embedded in an IFrame (no header is added), but when you set it to "false" the header is added
  and Airflow cannot be embedded in an IFrame.

New Features
^^^^^^^^^^^^

> - Add dynamic task mapping (`AIP-42 <https://github.com/apache/airflow/pulls?q=is%3Apr+is%3Amerged+label%3AAIP-42>`_)

Reviewer 5: This link is problematic.
If in the future we will add the AIP-42 label to a PR after this release then the PR will also be displayed in this list which will confuse users.
To solve this we must use milestone as part of the link and make sure all PRs that are relevant for this release are marked with the milestone

Reviewer 2: Good call. I set the milestone for all the PRs missing it.

```suggestion
- Add dynamic task mapping (`AIP-42 <https://github.com/apache/airflow/pulls?q=is%3Apr+is%3Amerged+label%3AAIP-42+milestone%3A%22Airflow+2.3.0%22>`_)
```

.. towncrier release notes start

Airflow 2.3.0 (2022-04-29)
--------------------------

> For production docker image related changes, see the `Docker Image Changelog <https://airflow.apache.org/docs/docker-stack/changelog.html>`_.

Reviewer 2: ```suggestion

For production docker image related changes, see the `Docker Image Changelog <https://airflow.apache.org/docs/docker-stack/changelog.html>`_.
```

.. towncrier release notes start

Airflow 2.3.0 (2022-04-29)
--------------------------

For production docker image related changes, see the `Docker Image Changelog <https://airflow.apache.org/docs/docker-stack/changelog.html>`_.

Significant Changes
^^^^^^^^^^^^^^^^^^^

Passing ``execution_date`` to ``XCom.set()``, ``XCom.clear()`` , ``XCom.get_one()`` , and ``XCom.get_many()`` is deprecated (#19825)

> """"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""

Reviewer 2: ```suggestion
""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""

```

Continuing the effort to bind TaskInstance to a DagRun, XCom entries are now also tied to a DagRun. Use the ``run_id`` argument to specify the DagRun instead.

Task log templates are now read from the metadata database instead of ``airflow.cfg`` (#20165)

> """"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""

Reviewer 2: ```suggestion
""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""

```

- retry commit on MySQL deadlocks during backfill (#22696)
- Add more fields to REST API get DAG(dags/dag_id) endpoint (#22637)
- Use timetable to generate planned days for current year (#22055)
- Disable connection pool for celery worker (#22493)
- Make date picker label visible in trigger dag view (#22379)
- Expose ``try_number`` in airflow vars (#22297)
- Add generic connection type (#22310)
- Add a few more fields to the taskinstance finished log message (#22262)
- Pause autorefresh if scheduler isn't running (#22151)
- Show DagModel details. (#21868)
- Add pip_install_options to PythonVirtualenvOperator (#22158)
- Show import error for ``airflow dags list`` CLI command (#21991)
- Pause auto-refresh when page is hidden (#21904)
- Default args type check (#21809)
- Enhance magic methods on XComArg for UX (#21882)
- py files don't have to be checked ``is_zipfiles`` in refresh_dag (#21926)
- Fix TaskDecorator type hints (#21881)
- Add 'Show record' option for variables (#21342)
- Use DB where possible for quicker ``airflow dag`` subcommands (#21793)
- REST API: add rendered fields in task instance. (#21741)
- Change the default auth backend to session (#21640)

> - Refactor ``TriggerRule`` & ``WeightRule`` classes to inherit from Enum (#21264)

Reviewer 5: Shouldn't this entry be in the internal section?

Reviewer 2: As these are meant to be end-user focused, we should probably just drop it, after looking at it again.

```suggestion
```

.. towncrier release notes start

Airflow 2.3.0 (2022-04-29)
--------------------------

Significant Changes
^^^^^^^^^^^^^^^^^^^

Passing ``execution_date`` to ``XCom.set()``\ , ``XCom.clear()``\ , ``XCom.get_one()``\ , and ``XCom.get_many()`` is deprecated
"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
Continuing the effort to bind TaskInstance to a DagRun, XCom entries are now also tied to a DagRun. Use the ``run_id`` argument to specify the DagRun instead. (#19825)

> Task log templates are now read from the metadatabase instead of ``airflow.cfg``

Reviewer 4: ```suggestion
Task log templates are now read from the metadata database instead of ``airflow.cfg``
```

""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
  Previously, a task’s log is dynamically rendered from the ``[core] log_filename_template`` and ``[elasticsearch] log_id_template`` config values at runtime. This resulted in unfortunate characteristics, e.g. it is impractical to modify the config value after an Airflow instance is running for a while, since all existing task logs have be saved under the previous format and cannot be found with the new config value.

>   A new ``log_template`` table is introduced to solve this problem. This table is synchronised with the aforementioned config values every time Airflow starts, and a new field ``log_template_id`` is added to every DAG run to point to the format used by tasks (\ ``NULL`` indicates the first ever entry for compatibility). (#20165)

Reviewer 4: ```suggestion
  A new ``log_template`` table is introduced to solve this problem. This table is synchronised with the aforementioned config values every time Airflow starts, and a new field ``log_template_id`` is added to every DAG run to point to the format used by tasks (``NULL`` indicates the first ever entry for compatibility). (#20165)
```

Minimum kubernetes version bumped from ``3.0.0`` to ``21.7.0``
""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""

  No change in behavior is expected.  This was necessary in order to take advantage of a `bugfix <https://github.com/kubernetes-client/python-base/commit/70b78cd8488068c014b6d762a0c8d358273865b4>`_ concerning refreshing of Kubernetes API tokens with EKS, which enabled the removal of some `workaround code <https://github.com/apache/airflow/pull/20759>`_. (#20759)

> XCom now define ``run_id`` instead of ``execution_date``

Reviewer 4: ```suggestion
XCom now defined by ``run_id`` instead of ``execution_date``
```

""""""""""""""""""""""""""""""""""""""""""""""""""""""""

>   As a continuation to the TaskInstance-DagRun relation change started in Airflow 2.2, the ``execution_date`` columns on XCom has been removed from the database, and replaced by an `association proxy <https://docs.sqlalchemy.org/en/13/orm/extensions/associationproxy.html>`_ field at the ORM level. If you access Airflow’s metadatabase directly, you should rewrite the implementation to use the ``run_id`` column instead.

Reviewer 4: ```suggestion
  As a continuation to the TaskInstance-DagRun relation change started in Airflow 2.2, the ``execution_date`` columns on XCom has been removed from the database, and replaced by an `association proxy <https://docs.sqlalchemy.org/en/13/orm/extensions/associationproxy.html>`_ field at the ORM level. If you access Airflow’s metadata database directly, you should rewrite the implementation to use the ``run_id`` column instead.
```

  Note that Airflow’s metadatabase definition on both the database and ORM levels are considered implementation detail without strict backward compatibility guarantees. (#20975)

Non-JSON-serializable params deprecated.
""""""""""""""""""""""""""""""""""""""""

  It was previously possible to use dag or task param defaults that were not JSON-serializable.

  For example this worked previously:

  .. code-block:: python

     @dag.task(params={"a": {1, 2, 3}, "b": pendulum.now()})
     def datetime_param(value):
         print(value)

     datetime_param("{{ params.a }} | {{ params.b }}")

  Note the use of ``set`` and ``datetime`` types, which are not JSON-serializable.  This behavior is problematic because to override these values in a dag run conf, you must use JSON, which could make these params non-overridable.  Another problem is that the support for param validation assumes JSON.  Use of non-JSON-serializable params will be removed in Airflow 3.0 and until then, use of them will produce a warning at parse time. (#21135)

You have to use ``postgresql://`` instead of ``postgres://`` in ``sql_alchemy_conn`` for SQLAlchemy 1.4.0+

> """"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""

Reviewer 4: ```suggestion
Must use ``postgresql://`` instead of ``postgres://`` in ``sql_alchemy_conn`` for SQLAlchemy 1.4.0+
""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
```

  When you use SQLAlchemy 1.4.0+, you need to use ``postgresql://`` as the scheme in the ``sql_alchemy_conn``.

>   In the previous versions of SQLAlchemy it was possible to use ``postgres://``\ , but using it in

Reviewer 4: ```suggestion
  In the previous versions of SQLAlchemy it was possible to use ``postgres://`` , but using it in
```

Deprecation: ``Connection.extra`` must be JSON-encoded dict
"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""

  TLDR
  ~~~~

  From Airflow 3.0, the ``extra`` field in airflow connections must be a JSON-encoded Python dict.

  What, why, and when?
  ~~~~~~~~~~~~~~~~~~~~

  Airflow's Connection is used for storing credentials.  For storage of information that does not
  fit into user / password / host / schema / port, we have the ``extra`` string field.  Its intention
  was always to provide for storage of arbitrary key-value pairs, like ``no_host_key_check`` in the SSH
  hook, or ``keyfile_dict`` in GCP.

  But since the field is string, it's technically been permissible to store any string value.  For example
  one could have stored the string value ``'my-website.com'`` and used this in the hook.  But this is a very

>   bad practice. One reason is intelligibility: when you look at the value for ``extra``\ , you don't have any idea

Reviewer 4: ```suggestion
  bad practice. One reason is intelligibility: when you look at the value for ``extra`` , you don't have any idea
```

The ``tree`` default view setting has been renamed to ``grid``
""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""

  If you set the ``dag_default_view`` config option or the ``default_view`` argument to ``DAG()`` to ``tree`` you will need to update your deployment. The old name will continue to work but will issue warnings. (#22167)

Database configuration moved to new section
"""""""""""""""""""""""""""""""""""""""""""

  The following configurations have been moved from ``[core]`` to the new ``[database]`` section. However when reading the new option, the old option will be checked to see if it exists. If it does a DeprecationWarning will be issued and the old option will be used instead.

  * sql_alchemy_conn
  * sql_engine_encoding
  * sql_engine_collation_for_ids
  * sql_alchemy_pool_enabled
  * sql_alchemy_pool_size
  * sql_alchemy_max_overflow
  * sql_alchemy_pool_recycle
  * sql_alchemy_pool_pre_ping
  * sql_alchemy_schema
  * sql_alchemy_connect_args
  * load_default_connections
  * max_db_retries (#22284)

Remove requirement that custom connection UI fields be prefixed
"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""

>   Hooks can define custom connection fields for their connection type by implementing method ``get_connection_form_widgets``.  These custom fields are presented in the web UI as additional connection attributes, but internally they are stored in the connection ``extra`` dict.  For technical reasons, previously custom field when stored in ``extra`` had to be named with a prefix ``extra__<conn type>__<field name>``.  This had the consequence of making it more cumbersome to define connections outside of the UI, since the prefix is tougher to read and work with. With #22607, we make it so that you can now define custom fields such that they can be read from and stored in ``extra`` without the prefix.

Reviewer 4: ```suggestion
  Hooks can define custom connection fields for their connection type by implementing method ``get_connection_form_widgets``.  These custom fields appear in the web UI as additional connection attributes, but internally they are stored in the connection ``extra`` dict field.  For technical reasons, previously, when stored in the ``extra`` dict, the custom field's dict key had to take the form ``extra__<conn type>__<field name>``.  This had the consequence of making it more cumbersome to define connections outside of the UI, since the prefix ``extra__<conn type>__`` makes it tougher to read and work with. With #22607, we make it so that you can now define custom fields such that they can be read from and stored in ``extra`` without the prefix.
```

>   To enable this, update the dict returned by the ``get_connection_form_widgets`` method to remove the prefix from the keys.  Internally, the providers manager will still use a prefix to ensure each custom field is globally unique, but the absence of a prefix in the returned widget dict will signal to the Web UI to read and store custom fields without the prefix.  Note that this is only a change to the Web UI behavior; when updating your hook in this way, you must make sure that when your *hook* reads the ``extra`` field, it will also check for the prefixed value for backward compatibility. (#22607)

Reviewer 4: ```suggestion
  To enable the use of short names in the extra dict, in the hook defining the custom fields you must update the dict returned by the ``get_connection_form_widgets`` method to remove the prefix from the keys.  Internally, the providers manager will still use a prefix to ensure each custom field is globally unique, but the absence of a prefix in the returned widget dict will signal to the Web UI to read and store custom fields without the prefix. Note that this is only a change to the Web UI behavior; when updating your hook in this way, you must make sure that when your *hook* reads the ``extra`` field, it will also check for the prefixed value for backward compatibility. (#22607)
```

  setting the value to "true" caused "X-Frame-Options" header to "DENY" (not allowing Airflow to be used
  in an IFrame). When you set it to "false", the header was not added, so Airflow could be embedded in an
  IFrame. By default Airflow could not be embedded in an IFrame.

  In Airflow 2.2.5 there was a bug introduced that made it impossible to disable Airflow to
  work in IFrame. No matter what the configuration was set, it was possible to embed Airflow in an IFrame.

  Airflow 2.3.0 restores the original meaning to the parameter. If you set it to "true" (default) Airflow
  can be embedded in an IFrame (no header is added), but when you set it to "false" the header is added
  and Airflow cannot be embedded in an IFrame. (#23222)

New Features
^^^^^^^^^^^^

- Add dynamic task mapping (`AIP-42 <https://github.com/apache/airflow/pulls?q=is%3Apr+is%3Amerged+label%3AAIP-42>`_)
- New Grid View replaces Tree View (#18675)
- Templated ``requirements.txt`` in Python Operators (#17349)
- Allow reuse of decorated tasks (#22941)
- Move the database configuration to a new section (#22284)
- Add ``SmoothOperator`` (#22813)

> - make operator's ``execution_timeout`` configurable (#22389)

Reviewer 4: ```suggestion
- Make operator's ``execution_timeout`` configurable (#22389)
```

- Events Timetable (#22332)
- Support dag serialization with custom ``ti_deps`` rules (#22698)
- Support log download in task log view (#22804)
- support for continue backfill on failures (#22697)
- Add ``dag-processor`` cli command (#22305)
- Add possibility to create users in LDAP mode (#22619)
- Add ``ignore_first_depends_on_past`` for scheduled jobs (#22491)
- Update base sensor operator to support XCOM return value (#20656)

> - add an option for run id in the ui trigger screen (#21851)

Reviewer 4: ```suggestion
- Add an option for run id in the ui trigger screen (#21851)
```

- Enable JSON serialization for connections (#19857)
- Add REST API endpoint for bulk update of DAGs (#19758)
- Add queue button to click-on-DagRun interface. (#21555)
- Add ``list-import-errors`` to ``airflow dags`` command (#22084)
- Store callbacks in database if ``standalone_dag_processor`` config is True. (#21731)

> - Local kubernetes executor (#19729)

Reviewer 2: ```suggestion
- Add LocalKubernetesExecutor (#19729)
```

> - add ``celery.task_timeout_error`` metric (#21602)

Reviewer 4: ```suggestion
- Add ``celery.task_timeout_error`` metric (#21602)
```

- Airflow ``db downgrade`` cli command (#21596)
- Add ``ALL_SKIPPED`` trigger rule (#21662)
- Add ``db clean`` CLI command for purging old data (#20838)
- Add ``celery_logging_level`` (#21506)
- Support different timeout value for dag file parsing (#21501)
- Support generating SQL script for upgrades (#20962)
- Add option to compress Serialized dag data (#21332)
- Branch python operator decorator (#20860)
- Add Audit Log View to Dag View (#20733)

> - Add missing statsd metric for failing SLA Callback notification (#20924)

Reviewer 2: ```suggestion
- Add missing StatsD metric for failing SLA Callback notification (#20924)
```

- Add option ``--skip-init`` to db reset command (#22989)
- Support importing connections from files with ".yml" extension (#22872)
- Support glob syntax in ``.airflowignore`` files (#21392) (#22051)
- Hide pagination when data is a single page (#22963)
- Support for sorting DAGs in the web UI (#22671)
- Speed up ``has_access`` decorator by ~200ms (#22858)
- Add XComArg to lazy-imported list of Airflow module (#22862)
- Add more fields to REST API dags/dag_id/details endpoint (#22756)
- Don't show irrelevant/duplicated/"internal" Task attrs in UI (#22812)
- No need to load whole ti in current_state (#22764)
- Pickle dag exception string fix (#22760)
- Better verification of Localexecutor's parallelism option (#22711)
- log backfill exceptions to sentry (#22704)
- retry commit on MySQL deadlocks during backfill (#22696)
- Add more fields to REST API get DAG(dags/dag_id) endpoint (#22637)
- Use timetable to generate planned days for current year (#22055)
- Disable connection pool for celery worker (#22493)
- Make date picker label visible in trigger dag view (#22379)
- Expose ``try_number`` in airflow vars (#22297)
- Add generic connection type (#22310)

> - add a few more fields to the taskinstance finished log message (#22262)

Reviewer 4: ```suggestion
- Add a few more fields to the taskinstance finished log message (#22262)
```

- Pause autorefresh if scheduler isn't running (#22151)
- Show DagModel details. (#21868)
- Add pip_install_options to PythonVirtualenvOperator (#22158)
- Show import error for ``airflow dags list`` CLI command (#21991)
- Pause auto-refresh when page is hidden (#21904)
- Default args type check (#21809)
- Enhance magic methods on XComArg for UX (#21882)

> - py files doesn't have to be checked ``is_zipfiles`` in refresh_dag (#21926)

Reviewer 4: ```suggestion
- py files don't have to be checked ``is_zipfiles`` in refresh_dag (#21926)
```

- Add docs for ``db upgrade`` / ``db downgrade`` (#21879)
- Update modules_management.rst (#21889)
- Fix UPDATING section on SqlAlchemy 1.4 scheme changes (#21887)
- Update TaskFlow tutorial doc to show how to pass "operator-level" args. (#21446)
- Fix doc - replace decreasing by increasing (#21805)
- Add another way to dynamically generate DAGs to docs (#21297)
- Add extra information about time synchronization needed (#21685)
- Update debug.rst docs (#21246)
- Replaces the usage of ``postgres://`` with ``postgresql://`` (#21205)
- Fix task execution process in ``CeleryExecutor`` docs (#20783)

Misc/Internal
"""""""""""""

- Bring back deprecated security manager functions (#23243)
- Replace usage of ``DummyOperator`` with ``EmptyOperator`` (#22974)
- Deprecate ``DummyOperator`` in favor of ``EmptyOperator`` (#22832)
- Remove unnecessary python 3.6 conditionals (#20549)
- Bump moment from 2.29.1 to 2.29.2 in /airflow/www (#22873)
- Bump prismjs from 1.26.0 to 1.27.0 in /airflow/www (#22823)
- Bump nanoid from 3.1.23 to 3.3.2 in /airflow/www (#22803)
- Bump minimist from 1.2.5 to 1.2.6 in /airflow/www (#22798)
- Remove dag parsing from db init command (#22531)

> - removed unnecessary dependence on bcrypt (#22498)

Reviewer 2: ```suggestion
```

This can be removed as it wasn't in a release.

- Update our approach for executor-bound dependencies (#22573)
- Use ``Airflow.Base.metadata`` in FAB models (#22353)

> - Update black precommit (#22521)

Reviewer 2: ```suggestion
```

> - New design of system tests (#22311)

Reviewer 2: ```suggestion
```

I'm not sure we need this in the release notes.

- Limit Docutils to make our documentation pretty again (#22420)
- Add Python 3.10 support (#22050)
- [FEATURE] add 1.22 1.23 K8S support (#21902)
- Remove pandas upper limit now that SQLA is 1.4+ (#22162)
- Patch ``sql_alchemy_conn`` if old postgres scheme used (#22333)
- Protect against accidental misuse of XCom.get_value() (#22244)

> - Remove Snowflake limits (#22181)

Reviewer 2: ```suggestion
```

This is a provider change.

- Order filenames for migrations (#22168)
- Don't try to create automigration for celery tables (#22120)
- Require SQLAlchemy 1.4 (#22114)

> - bump moto version (#22099)

Reviewer 2: ```suggestion
```

This is devel only.

- bump sphinx-jinja (#22101)

> - Upgrade ``moto`` library to version 3.0 (#22005)

Reviewer 2: ```suggestion
```

> - Simplify tests for CLI connections commands (#21983)

Reviewer 2: ```suggestion
```

> - Add test to run DB downgrade in the CI (#21273)

Reviewer 2: ```suggestion
```

- Add compat shim for SQLAlchemy to avoid warnings (#21959)
- Rename ``xcom.dagrun_id`` to ``xcom.dag_run_id`` (#21806)
- Deprecate non-JSON ``conn.extra`` (#21816)

> - Autogenerate migration reference doc (#21601)

Reviewer 2: ```suggestion
```

> - Limits GitHub3.py in order to avoid backtracking (#21824)

Reviewer 2: ```suggestion
```

- Bump upper bound version of ``jsonschema`` to 5.0 (#21712)
- Deprecate helper utility ``days_ago`` (#21653)
- Remove ```:type``` lines now sphinx-autoapi supports typehints (#20951)
- Quelch deprecation warning in tests (#20900)
- Use ``DagRun.run_id`` instead of ``execution_date`` when updating state of TIs(UI & REST API) (#18724)
- Add Context stub to Airflow packages (#20817)
- Update Kubernetes library version (#18797)
- Rename ``PodLauncher`` to ``PodManager`` (#20576)
- Removes Python 3.6 support (#20467)
- Add deprecation warning for non-json-serializable params (#20174)
- Rename TaskMixin to DependencyMixin (#20297)
- Deprecate passing execution_date to XCom methods (#19825)
- Remove ``get_readable_dags`` and ``get_editable_dags``, and ``get_accessible_dags``. (#19961)
- Remove postgres 9.6 support (#19987)
- Removed hardcoded connection types. Check if hook is instance of DbApiHook. (#19639)
- add kubernetes 1.21 support (#19557)
- Add FAB base class and set import_name explicitly. (#19667)

> - Remove remaining ``pylint: disable`` comments (#19541)

Reviewer 2: ```suggestion
```

- Removes unused state transitions to handle auto-changing view permissions. (#19153)
- Chore: Use enum for ``__var`` and ``__type`` members (#19303)
- Use fab models (#19121)
- Consolidate method names between Airflow Security Manager and FAB default (#18726)

> - Fixup string concatenations (#19099)

Reviewer 2: ```suggestion
```

- Remove distutils usages for Python 3.10 (#19064)
- Removing redundant ``max_tis_per_query`` initialisation on SchedulerJob (#19020)

> - More f-strings (#18855)

Reviewer 2: ```suggestion
```

> - Add pre-commit hook for common misspelling check in files (#18964)

Reviewer 2: ```suggestion
```

- Remove deprecated usage of ``init_role()`` from API (#18820)

> - Ignore License check in ``airflow/www/static/robots.txt`` (#18886)

Reviewer 2: ```suggestion
```
